diff -Naur linux-2.6.24.7/include/linux/sched.h linux-2.6.24.7_mod/include/linux/sched.h
--- linux-2.6.24.7/include/linux/sched.h	2008-05-06 16:22:34.000000000 -0700
+++ linux-2.6.24.7_mod/include/linux/sched.h	2008-10-07 16:58:11.000000000 -0700
@@ -1178,6 +1178,12 @@
 	int make_it_fail;
 #endif
 	struct prop_local_single dirties;
+
+	/* My stuff */
+	u64 interval;
+	u64 inst;
+	u64 re_cy;
+	u64 ref_cy;
 };
 
