diff -Naur linux-2.6.25.18/include/linux/sched.h linux-2.6.25.18_Sched_Mod/include/linux/sched.h
--- linux-2.6.25.18/include/linux/sched.h	2008-10-08 19:58:32.000000000 -0700
+++ linux-2.6.25.18_Sched_Mod/include/linux/sched.h	2008-10-11 16:06:14.000000000 -0700
@@ -1,6 +1,8 @@
 #ifndef _LINUX_SCHED_H
 #define _LINUX_SCHED_H
 
+#define SEEKER_PLUGIN_PATCH 1
+
 /*
  * cloning flags:
  */
@@ -1271,6 +1273,13 @@
 	int latency_record_count;
 	struct latency_record latency_record[LT_SAVECOUNT];
 #endif
+
+	/* Seeker Patch Additions */
+	u64 interval;
+	u64 inst;
+	u64 re_cy;
+	u64 ref_cy;
+
 };
 
 /*
