diff -Naur linux-2.6.25.18/include/linux/sched.h linux-2.6.25.18_sched/include/linux/sched.h
--- linux-2.6.25.18/include/linux/sched.h	2008-10-08 19:58:32.000000000 -0700
+++ linux-2.6.25.18_sched/include/linux/sched.h	2008-10-19 02:37:42.000000000 -0700
@@ -1,6 +1,8 @@
 #ifndef _LINUX_SCHED_H
 #define _LINUX_SCHED_H
 
+#define SEEKER_PLUGIN_PATCH 1
+
 /*
  * cloning flags:
  */
@@ -1271,6 +1273,13 @@
 	int latency_record_count;
 	struct latency_record latency_record[LT_SAVECOUNT];
 #endif
+
+	/* Seeker Patch Additions */
+	u64 interval;
+	short cpustate;
+	u64 inst;
+	u64 re_cy;
+	u64 ref_cy;
+
 };
 
 /*
diff -Naur linux-2.6.25.18/kernel/sched.c linux-2.6.25.18_sched/kernel/sched.c
--- linux-2.6.25.18/kernel/sched.c	2008-10-08 19:58:32.000000000 -0700
+++ linux-2.6.25.18_sched/kernel/sched.c	2008-10-19 02:34:07.000000000 -0700
@@ -1461,6 +1461,7 @@
 {
 	return cpu_rq(cpu)->load.weight;
 }
+EXPORT_SYMBOL_GPL(weighted_cpuload);
 
 static inline void __set_task_cpu(struct task_struct *p, unsigned int cpu)
 {
